ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -1303,3 +1338,223 @@ bool attribute_is_exportable(struct pkcs11_attribute_head *req_attr,
 
 	return true;
 }
+
+static bool attr_is_modifiable_any_key(struct pkcs11_attribute_head *attr)
+{
+	switch (attr->id) {
+	case PKCS11_CKA_ID:
+	case PKCS11_CKA_START_DATE:
+	case PKCS11_CKA_END_DATE:
+	case PKCS11_CKA_DERIVE:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static bool attr_is_modifiable_secret_key(struct pkcs11_attribute_head *attr,
+					  struct pkcs11_session *session,
+					  struct pkcs11_object *obj)
+{
+	switch (attr->id) {
+	case PKCS11_CKA_ENCRYPT:
+	case PKCS11_CKA_DECRYPT:
+	case PKCS11_CKA_SIGN:
+	case PKCS11_CKA_VERIFY:
+	case PKCS11_CKA_WRAP:
+	case PKCS11_CKA_UNWRAP:
+		return true;
+	/* Can't be modified once set to CK_FALSE - 12 in Table 10 */
+	case PKCS11_CKA_EXTRACTABLE:
+		return get_bool(obj->attributes, PKCS11_CKA_EXTRACTABLE);
+	/* Can't be modified once set to CK_TRUE - 11 in Table 10 */
+	case PKCS11_CKA_SENSITIVE:
+	case PKCS11_CKA_WRAP_WITH_TRUSTED:
+		return !get_bool(obj->attributes, attr->id);
+	/* Change in CKS_TRUSTED can only be done by SO */
+	case PKCS11_CKA_TRUSTED:
+		return pkcs11_session_is_so(session);
+	default:
+		return false;
+	}
+}
+
+static bool attr_is_modifiable_public_key(struct pkcs11_attribute_head *attr,
+					  struct pkcs11_session *session,
+					  struct pkcs11_object *obj __unused)
+{
+	switch (attr->id) {
+	case PKCS11_CKA_SUBJECT:
+	case PKCS11_CKA_ENCRYPT:
+	case PKCS11_CKA_VERIFY:
+	case PKCS11_CKA_VERIFY_RECOVER:
+	case PKCS11_CKA_WRAP:
+		return true;
+	case PKCS11_CKA_TRUSTED:
+		/* Change in CKS_TRUSTED can only be done by SO */
+		return pkcs11_session_is_so(session);
+	default:
+		return false;
+	}
+}
+
+static bool attr_is_modifiable_private_key(struct pkcs11_attribute_head *attr,
+					   struct pkcs11_session *sess __unused,
+					   struct pkcs11_object *obj)
+{
+	switch (attr->id) {
+	case PKCS11_CKA_SUBJECT:
+	case PKCS11_CKA_DECRYPT:
+	case PKCS11_CKA_SIGN:
+	case PKCS11_CKA_SIGN_RECOVER:
+	case PKCS11_CKA_UNWRAP:
+	/*
+	 * TBD: Revisit if we don't support PKCS11_CKA_PUBLIC_KEY_INFO
+	 * Specification mentions that if this attribute is
+	 * supplied as part of a template for C_CreateObject, C_CopyObject or
+	 * C_SetAttributeValue for a private key, the token MUST verify
+	 * correspondence between the private key data and the public key data
+	 * as supplied in CKA_PUBLIC_KEY_INFO. This needs to be
+	 * taken care of when this object type will be implemented
+	 */
+	case PKCS11_CKA_PUBLIC_KEY_INFO:
+		return true;
+	/* Can't be modified once set to CK_FALSE - 12 in Table 10 */
+	case PKCS11_CKA_EXTRACTABLE:
+		return get_bool(obj->attributes, PKCS11_CKA_EXTRACTABLE);
+	/* Can't be modified once set to CK_TRUE - 11 in Table 10 */
+	case PKCS11_CKA_SENSITIVE:
+		return !get_bool(obj->attributes, PKCS11_CKA_SENSITIVE);
+	case PKCS11_CKA_WRAP_WITH_TRUSTED:
+		return !get_bool(obj->attributes, PKCS11_CKA_WRAP_WITH_TRUSTED);
+	default:
```

Comment:
return false for:

- CKA_NEVER_EXTRACTABLE
- CKA_ALWAYS_SENSITIVE

Reply:
Same here. Any attribute not explicitly specified results in a return of false.